src/data/enum.lean

```diff
@@ -0,0 +1,190 @@
+
+import category.monad.basic
+import data.list.basic
+import data.equiv.basic
+import data.finset
+import data.fintype
+
+open finset (hiding singleton)
+
+namespace fin
+
+def enum (n : ℕ) : list (fin n) :=
+(list.range n).attach.map $ λ ⟨i,h⟩, ⟨i,by simpa using h⟩
+
+@[simp] lemma mem_enum {n} (x : fin n) : x ∈ enum n :=
+by simp [enum]; existsi x.1; simp [fin.is_lt]
+
+end fin
+
+class enumerable (α : Sort*) :=
+(card : ℕ)
+(equiv : α ≃ fin card)
+
+namespace enumerable
+
+variables {α : Type*}
+
+instance [enumerable α] : decidable_eq α :=
```

Comment:
Is this instance necessary? I'm concerned that in some cases (for example, the instance `enumerable (α × β)` below) this `decidable_eq` instance could be a lot less efficient than the obvious one (in this case, checking for equality of each component separately).

Reply:
I could make it a level 10 instance. Or I could make it a component of the class and give it a default definition.